File: Abyss Project/AreaCount.cpp
bool is_complete(bool[][] _input_array){

for(int i=0; i < y_dim; i++){
    for(int j=0; j < x_dim; j++){

        if(_input_array[i][j] == false)
            return false;
    }
   return true;
 }
}

def flood_fill(x, y, old, _input_array, _shape):

    to_fill = set()
    to_fill.add((x, y))
    while len(to_fill) != 0:
        (x, y) = to_fill.pop()

        if(x < _shape[0] and y < _shape[1]) and (_input_array[x][y] == old):
            _input_array[x][y] = True
            to_fill.add((x+1, y))
            to_fill.add((x, y-1))
            to_fill.add((x-1, y))
            to_fill.add((x, y+1))

void flood_fill(int x, int y, int old, int[][] _input_array, int[] _shape){

    set <int, greater <int> > to_fill;
    to_fill.insert((x, y));
    while(to_fill.size != 0){

        (x, y) = to_fill.erase((x, y));

        if(((x < _shape[0]) & (y < _shape[1])) & (_input_array[x][y] == old){

            _input_array[x][y] = -1;
            to_fill.insert((x+1, y))
            to_fill.insert((x, y-1))
            to_fill.insert((x-1, y))
            to_fill.insert((x, y+1))

           }

    }

}

File: Abyss Project/areacount.py
import numpy as np
import sys
# import cv2
# import time

np.set_printoptions(threshold=sys.maxsize)
sample = np.fromfile('sample.bin', dtype='uint8')
sample = sample.reshape(256, 256)
a = [0] * 256


def flood_fill(x, y, old):

    toFill = set()
    toFill.add((x, y))
    while len(toFill) != 0:
        (x, y) = toFill.pop()

        # cv2.imwrite('sample2.png', sample)
        if(x < 256 and y < 256) and (sample[x][y] == old):
            sample[x][y] = True
            toFill.add((x+1, y))
            toFill.add((x-1, y))
            toFill.add((x, y+1))
            toFill.add((x, y-1))


u = np.unique(sample)


def get_shade_index(shade):
    for row in range(256):
        for col in range(256):
            if(sample[row][col] == shade):
                return (row, col)
    return (None, None)


def is_complete():
    for row in sample:
        for x in row:
            if(x != True):
                return False
    return True


while(not(is_complete())):
    for shade in u:
        row,col = get_shade_index(shade)
        if( row != None ):
            flood_fill(row, col, shade)
            a[shade] = a[shade] + 1


print(a)
print('Black: ', a[0])
print('Grey: ', a[200])
print('white: ', a[255])
# cv2.imwrite('sample2.png', sample)








